File: metadata/mill-metadata-autoconfigure/src/main/java/io/qpointz/mill/metadata/repository/file/SpringResourceResolver.java
package io.qpointz.mill.metadata.repository.file;

import org.springframework.core.io.Resource;
import org.springframework.core.io.ResourceLoader;
import org.springframework.core.io.support.ResourcePatternResolver;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * Spring-based implementation of {@link ResourceResolver}.
 * Supports glob patterns via {@link ResourcePatternResolver} and simple paths via {@link ResourceLoader}.
 */
public class SpringResourceResolver implements ResourceResolver {

    private final ResourceLoader resourceLoader;

    public SpringResourceResolver(ResourceLoader resourceLoader) {
        this.resourceLoader = resourceLoader;
    }

    @Override
    public List<ResolvedResource> resolve(String locationPattern) throws IOException {
        List<ResolvedResource> result = new ArrayList<>();

        ResourcePatternResolver patternResolver = null;
        if (resourceLoader instanceof ResourcePatternResolver) {
            patternResolver = (ResourcePatternResolver) resourceLoader;
        }

        if (patternResolver != null && (locationPattern.contains("*") || locationPattern.contains("?"))) {
            Resource[] matched = patternResolver.getResources(locationPattern);
            for (Resource resource : matched) {
                if (resource.exists()) {
                    result.add(new ResolvedResource(resource.toString(), resource.getInputStream()));
                }
            }
        } else {
            Resource resource = resourceLoader.getResource(locationPattern);
            if (resource.exists()) {
                result.add(new ResolvedResource(resource.toString(), resource.getInputStream()));
            }
        }

        return result;
    }
}

File: metadata/mill-metadata-autoconfigure/src/main/kotlin/io/qpointz/mill/metadata/repository/file/SpringResourceResolver.kt
package io.qpointz.mill.metadata.repository.file

import org.springframework.core.io.ResourceLoader
import org.springframework.core.io.support.ResourcePatternResolver

/** Spring-based [ResourceResolver] supporting plain paths and glob patterns. */
class SpringResourceResolver(private val resourceLoader: ResourceLoader) : ResourceResolver {

    override fun resolve(locationPattern: String): List<ResolvedResource> {
        val result = mutableListOf<ResolvedResource>()
        val patternResolver = resourceLoader as? ResourcePatternResolver

        if (patternResolver != null && (locationPattern.contains("*") || locationPattern.contains("?"))) {
            val matched = patternResolver.getResources(locationPattern)
            for (resource in matched) {
                if (resource.exists()) {
                    result.add(ResolvedResource(resource.toString(), resource.inputStream))
                }
            }
        } else {
            val resource = resourceLoader.getResource(locationPattern)
            if (resource.exists()) {
                result.add(ResolvedResource(resource.toString(), resource.inputStream))
            }
        }
        return result
    }
}
